# Group and Join Practice

In [2]:
# Just run this cell
import warnings
warnings.filterwarnings("ignore", message="The 'warn' parameter of")

from datascience import *
import numpy as np

The following cells will load in 3 different tables:

## `salesperson`
contains information about the different salespersons at a company including their ID number, Name, City in which they work, and the percent comission they make on any sale they make.

In [3]:
salesperson = Table(['Salesperson ID', 'Salesperson Name', 'City', 'Commission']).with_rows([
    ['5001', 'James Hoog',  'New York', 0.15],
    ['5002', 'Nail Knite',  'Paris',    0.13],
    ['5005', 'Amoli Alex',  'London',   0.11],
    ['5006', 'Mike Lyon',   'Paris',    0.14],
    ['5007', 'Adam Plank',  'Rome',     0.13],
    ['5003', 'Layla Ahmad', 'San Jose', 0.12]
])

salesperson

Salesperson ID | Salesperson Name | City     | Commission
5001           | James Hoog       | New York | 0.15
5002           | Nail Knite       | Paris    | 0.13
5005           | Amoli Alex       | London   | 0.11
5006           | Mike Lyon        | Paris    | 0.14
5007           | Adam Plank       | Rome     | 0.13
5003           | Layla Ahmad      | San Jose | 0.12

## `customer`

This table contains customer information, including their ID number, Name, City, their grade (a number that indicates the volume of purchases they make), and their assigned salesperson indicated by the salesperson ID number.

In [4]:
customer = Table(['Customer ID', 'Customer Name', 'City', 'Grade', 'Salesperson ID']).with_rows([
    ['3002', 'Nick Rimando',   'New York',   100, '5001'],
    ['3007', 'Brad Davis',     'New York',   200, '5001'],
    ['3005', 'Graham Zusi',    'California', 200, '5002'],
    ['3008', 'Jilian Green',   'London',     300, '5002'],
    ['3004', 'Fabian Johnson', 'Paris',      300, '5006'],
    ['3009', 'Gina Cameron',   'Berlin',     100, '5003'],
    ['3003', 'Helen Li',       'Moscow',     200, '5007'],
    ['3001', 'Corey Jackson',  'London',     200, '5005']
])

customer

Customer ID | Customer Name  | City       | Grade | Salesperson ID
3002        | Nick Rimando   | New York   | 100   | 5001
3007        | Brad Davis     | New York   | 200   | 5001
3005        | Graham Zusi    | California | 200   | 5002
3008        | Jilian Green   | London     | 300   | 5002
3004        | Fabian Johnson | Paris      | 300   | 5006
3009        | Gina Cameron   | Berlin     | 100   | 5003
3003        | Helen Li       | Moscow     | 200   | 5007
3001        | Corey Jackson  | London     | 200   | 5005

## `orders`

This table contains information about recent orders, including the Order Number, the purchase amount (in dollars), the date of the order, the customer who made the purchase (indicated by their ID number) and the salesperson who made the sale (indicated by ID number).

In [5]:
orders = Table(['Order Number', 'Purchase Amount', 'Order Date', 'Customer ID', 'Salesperson ID']).with_rows([
    ['70001', '$150.50',   '2019-10-05', '3005', '5002'],
    ['70009', '$270.65',   '2019-09-10', '3001', '5005'],
    ['70002', '$65.26',    '2019-10-05', '3002', '5001'],
    ['70004', '$110.50',   '2019-08-17', '3009', '5003'],
    ['70007', '$948.50',   '2019-09-10', '3005', '5002'],
    ['70005', '$2,400.60', '2019-07-27', '3007', '5001'],
    ['70008', '$5,760.00', '2019-09-10', '3002', '5001'],
    ['70010', '$1,983.43', '2019-10-10', '3004', '5006'],
    ['70003', '$2,480.40', '2019-10-10', '3009', '5003'],
    ['70012', '$250.45',   '2019-06-27', '3008', '5002'],
    ['70011', '$75.29',    '2019-08-17', '3003', '5007'],
    ['70013', '$3,045.60', '2019-04-25', '3002', '5001']
])

orders

Order Number | Purchase Amount | Order Date | Customer ID | Salesperson ID
70001        | $150.50         | 2019-10-05 | 3005        | 5002
70009        | $270.65         | 2019-09-10 | 3001        | 5005
70002        | $65.26          | 2019-10-05 | 3002        | 5001
70004        | $110.50         | 2019-08-17 | 3009        | 5003
70007        | $948.50         | 2019-09-10 | 3005        | 5002
70005        | $2,400.60       | 2019-07-27 | 3007        | 5001
70008        | $5,760.00       | 2019-09-10 | 3002        | 5001
70010        | $1,983.43       | 2019-10-10 | 3004        | 5006
70003        | $2,480.40       | 2019-10-10 | 3009        | 5003
70012        | $250.45         | 2019-06-27 | 3008        | 5002
... (2 rows omitted)

### Question 1

Use the `join` and `select` methods to create a Table in which each row details a salesperson and customer pairing and the city in which the salesperson resides. For example:

|Salesperson Name|Customer Name|City|
|----------------|-------------|----|
|James Hoog|Nick Rimando|New York|
|James Hoog|Brad Davis|New York|
|Nail Knite|Graham Zusi|Paris|
|Nail Knite|Jilian Green|Paris|
|Layla Ahmad|Gina Cameron|San Jose|
|Amoli Alex|Corey Jackson|London|
|Mike Lyon|Fabian Johnson|Paris|
|Adam Plank|Helen Li|Rome|

### Question 2

Use the `join` method to make the table below that details the order information:

|Order Number|Purchase Amount|Customer Name|City|
|------------|---------------|-------------|----|
|70001|\$150.50|Graham Zusi|California|
|70002|\$65.26|Nick Rimando|New York|
|70003|\$2,480.40|Gina Cameron|Berlin|
|70004|\$110.50|Gina Cameron|Berlin|
|70005|\$2,400.60|Brad Davis|New York|
|70007|\$948.50|Graham Zusi|California|
|70008|\$5,760.00|Nick Rimando|New York|
|70009|\$270.65|Corey Jackson|London|
|70010|\$1,983.43|Fabian Johnson|Paris|
|70011|\$75.29|Helen Li|Moscow|
|70012|\$250.45|Jilian Green|London|
|70013|\$3,045.60|Nick Rimando|New York|

### Question 3

Use the `join` method to make the table below:

|Customer Name|City|Salesperson Name|Commission|
|-------------|----|----------------|----------|
|Nick Rimando|New York|James Hoog|0.15|
|Brad Davis|New York|James Hoog|0.15|
|Graham Zusi|California|Nail Knite|0.13|
|Jilian Green|London|Nail Knite|0.13|
|Gina Cameron|Berlin|Layla Ahmad|0.12|
|Corey Jackson|London|Amoli Alex|0.11|
|Fabian Johnson|Paris|Mike Lyon|0.14|
|Helen Li|Moscow|Adam Plank|0.13|

### Question 4

Use the `join` method to make the table below:

|Salesperson Name|Customer Name|City|Commission|
|----------------|-------------|----|----------|
|James Hoog|Nick Rimando|New York|0.15|
|James Hoog|Brad Davis|New York|0.15|
|Nail Knite|Graham Zusi|Paris|0.13|
|Nail Knite|Jilian Green|Paris|0.13|
|Mike Lyon|Fabian Johnson|Paris|0.14|
|Adam Plank|Helen Li|Rome|0.13|


**Note:** Only display the customers whose appointed salesperson gets a commission of more than 12%.

### Question 5

Use the `join` method to make the table below:

|Customer Name|Customer City|Salesperson Name|Salesperson City|Commission|
|-------------|----|----------------|------|----------|
|Graham Zusi|Paris|Nail Knite|California|0.13|
|Jilian Green|Paris|Nail Knite|London|0.13|
|Gina Cameron|San Jose|Layla Ahmad|Berlin|0.12|
|Helen Li|Rome|Adam Plank|Moscow|0.13|

**Note 1:** Only display the customers whose appointed salesperson lives in a different city.

**Note 2:** Relabel the column names to `Customer City` and `Salesperson City`.

**Hint 1:** The `.relabled` method allows you to relabel multiple comluns using arrays. For more information on the `.relabled` method [click here](http://data8.org/datascience/_autosummary/datascience.tables.Table.relabeled.html) to look at the documentation.

### Question 6

Use the `join` method to make the table below:

|Order Number|Order Date|Purchase Amount|Customer Name|Grade|Salesperson Name|Commission|
|------------|----------|---------------|-------------|-----|----------------|----------|
|70002|2019-10-05|\$65.26|Nick Rimando|100|James Hoog|0.15|
|70008|2019-09-10|\$5,760.00|Nick Rimando|100|James Hoog|0.15|
|70013|2019-04-25|\$3,045.60|Nick Rimando|100|James Hoog|0.15|
|70005|2019-07-27|\$2,400.60|Brad Davis|200|James Hoog|0.15|
|70001|2019-10-05|\$150.50|Graham Zusi|200|Nail Knite|0.13|
|70007|2019-09-10|\$948.50|Graham Zusi|200|Nail Knite|0.13|
|70012|2019-06-27|\$250.45|Jilian Green|300|Nail Knite|0.13|
|70004|2019-08-17|\$110.50|Gina Cameron|100|Layla Ahmad|0.12|
|70003|2019-10-10|\$2,480.40|Gina Cameron|100|Layla Ahmad|0.12|
|70009|2019-09-10|\$270.65|Corey Jackson|200|Amoli Alex|0.11|

### Question 7

Use the `join` method to make the table below:

|Order Number|Order Date|Purchase Amount|Customer Name|Grade|Salesperson Name|Commission|Commission Earned|
|------------|----------|---------------|-------------|-----|----------------|----------|-----------------|
|70002|2019-10-05|\$65.26|Nick Rimando|100|James Hoog|0.15|\$9.79|
|70008|2019-09-10|\$5,760.00|Nick Rimando|100|James Hoog|0.15|\$864.00|
|70013|2019-04-25|\$3,045.60|Nick Rimando|100|James Hoog|0.15|\$456.84|
|70005|2019-07-27|\$2,400.60|Brad Davis|200|James Hoog|0.15|\$360.09|
|70001|2019-10-05|\$150.50|Graham Zusi|200|Nail Knite|0.13|\$19.57|
|70007|2019-09-10|\$948.50|Graham Zusi|200|Nail Knite|0.13|\$123.30|
|70012|2019-06-27|\$250.45|Jilian Green|300|Nail Knite|0.13|\$32.56|
|70004|2019-08-17|\$110.50|Gina Cameron|100|Layla Ahmad|0.12|\$13.26|
|70003|2019-10-10|\$2,480.40|Gina Cameron|100|Layla Ahmad|0.12|\$297.65|
|70009|2019-09-10|\$270.65|Corey Jackson|200|Amoli Alex|0.11|\$29.77|
|70010|2019-10-10|\$1,983.43|Fabian Johnson|300|Mike Lyon|0.14|\$277.68|
|70011|2019-08-17|\$75.29|Helen Li|200|Adam Plank|0.13|\$9.79|

**Note 1:** You will need functions to convert the `Purchase Amount` from a string to a number and to calculate the `Commission Earned`.

**Hint 1:** You can use the `.apply` method to call a function on each item in an array.

**Hint 2:** The `.replace` method to remove characters from any position in a string. For more information on the `.replace` method [click here](http://data8.org/fa19/python-reference.html) to look at the documentation.

**Hint 3:** A function can have multiple arguments. Define a function that accepts the purchase amount and the commission percent, then calculates the commission earned.

**Hint 4:** The `CurrencyFormatter` will display a `float` with the '\$' symbol. For more information on `CurrencyFormatter` [click here](http://data8.org/datascience/formats.html) to look at the documentation. 

## Optional Challenge

### Question 8

Use the `join` method to make the table below:

|Salesperson Name|Total Sales|
|----------------|-----------|
|Adam Plank|\$75|
|Amoli Alex|\$271|
|James Hoog|\$11,271|
|Layla Ahmad|\$2,591|
|Mike Lyon|\$1,983|
|Nail Knite|\$1,349|

**Hint 1:** Use the `np.unique()` function to return the unique elements from the `Salesperson Name` column.